crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -702,19 +784,336 @@ impl<'src> CodeExampleDoctest<'src> {
+            None => {
+                queue.push_back(self.into_format_action());
+                return None;
+            }
             Some(code) => code,
         };
         self.lines.push(CodeExampleLine { original, code });
-        true
+        queue.push_back(CodeExampleAddAction::Kept);
+        Some(self)
+    }
+
+    /// Consume this doctest and turn it into a formatting action.
+    fn into_format_action(self) -> CodeExampleAddAction<'src> {
+        CodeExampleAddAction::Format {
+            kind: CodeExampleKind::Doctest(self),
+        }
+    }
+}
+
+/// State corresponding to a single reStructuredText literal block or
+/// code-block directive.
+///
+/// While a literal block and code-block directive are technically two
+/// different reStructuredText constructs, we use one type to represent
+/// both because they are exceptionally similar. Basically, they are
+/// the same with two main differences:
+///
+/// 1. Literal blocks are began with a line that ends with `::`. Code block
+/// directives are began with a line like `.. code-block:: python`.
+/// 2. Code block directives permit a list of options as a "field list"
+/// immediately after the opening line. Literal blocks have no options.
+///
+/// Otherwise, everything else, including the indentation structure, is the
+/// same.
+#[derive(Debug)]
+struct CodeExampleRst<'src> {
+    /// The lines that have been seen so far that make up the block.
+    lines: Vec<CodeExampleLine<'src>>,
+    /// The indent of the line "opening" this block. It can either be the
+    /// indent of a line ending with `::` (for a literal block) or the indent
+    /// of a line starting with `.. ` (a directive).
+    ///
+    /// The content body of a block needs to be indented more than the line
+    /// opening the block, so we use this indentation to look for indentation
+    /// that is "more than" it.
+    opening_indent: TextSize,
+    /// The minimum indent of the block.
+    ///
+    /// This is `None` until the first such line is seen. If no such line is
+    /// found, then we consider it an invalid block and bail out of trying to
+    /// find a code snippet. Otherwise, we update this indentation as we see
+    /// lines in the block with less indentation. (Usually, the minimum is the
+    /// indentation of the first block, but this is not required.)
+    ///
+    /// By construction, all lines part of the block must have at least this
+    /// indentation. Additionally, it is guaranteed that the indentation length
+    /// of the opening indent is strictly less than the indentation of the
+    /// minimum indent. Namely, the block ends once we find a line that has
+    /// been unindented to at most the indent of the opening line.
+    ///
+    /// When the code snippet has been extracted, it is re-built before being
+    /// reformatted. The minimum indent is stripped from each line when it is
+    /// re-built.
+    min_indent: Option<TextSize>,
```

Comment:
What's the unit in which the `indent` is measured? Is it the byte length of the indentation string, meaning that `\t` and ` ` space both count as one?

Reply:
No, it's computed by `indentation_length`:

```rust
/// For docstring indentation, black counts spaces as 1 and tabs by increasing the indentation up
/// to the next multiple of 8. This is effectively a port of
/// [`str.expandtabs`](https://docs.python.org/3/library/stdtypes.html#str.expandtabs),
/// which black [calls with the default tab width of 8](https://github.com/psf/black/blob/c36e468794f9256d5e922c399240d49782ba04f1/src/black/strings.py#L61).
fn indentation_length(line: &str) -> TextSize {
    let mut indentation = 0u32;
    for char in line.chars() {
        if char == '\t' {
            // Pad to the next multiple of tab_width
            indentation += 8 - (indentation.rem_euclid(8));
        } else if char.is_whitespace() {
            indentation += u32::from(char.text_len());
        } else {
            break;
        }
    }
    TextSize::new(indentation)
}
```

I added a clarifying comment to the docs here.